### PR TITLE
Refactor SignerSource to expose DerivationPath to other kinds of signers

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
DerivationPath parsing doesn't lend itself to code sharing (only available for `SignerSource::Usb`-type signers, within remote-wallet).

#### Summary of Changes
Refactor it:
- Add `DerivationPath::from_uri()` method that implements the DerivationPath parsing previously in `Locator::new_from_uri()`; move test cases to target derivation-path specifically
- Convert `SignerSource` to a struct that wraps the old `kind` and includes and optional `derivation_path`

Toward #5246

### CODE REVIEW

File: clap-utils/src/keypair.rs [hunk 1]
```diff
@@ -142,27 +143,35 @@ pub(crate) enum SignerSource {
...
             if let Some(scheme) = uri.scheme() {
                 let scheme = scheme.as_str().to_ascii_lowercase();
                 match scheme.as_str() {
-                    "ask" => SignerSource::Ask,
-                    "file" => SignerSource::Filepath(uri.path().to_string()),
-                    "stdin" => SignerSource::Stdin,
-                    "usb" => SignerSource::Usb(source.to_string()),
-                    _ => SignerSource::Filepath(source.to_string()),
+                    "ask" => Ok(SignerSource::Ask),
+                    "file" => Ok(SignerSource::Filepath(uri.path().to_string())),
```
Comment (t-nelson) [inline clap-utils/src/keypair.rs]: wdyt about `stat()`ing the path here and failing if it doesn't exist with readable permissions?

**Review by t-nelson — COMMENTED** [review]
Looking great!  Much cleaner.  Just a few suggestsions

File: sdk/src/derivation_path.rs [hunk 2]
```diff
@@ -120,6 +122,31 @@ impl DerivationPath {
...
+            let key = query.get("key");
+            if key.is_none() {
+                return Err(DerivationPathError::InvalidDerivationPath(format!(
+                    "invalid query string `{}`, only `key` supported",
+                    query_str,
+                )));
+            }
+            // Use from_key_str instead of TryInto here to make it a little more explicit that this
+            // generates a Solana bip44 DerivationPath
+            Self::from_key_str(key.unwrap()).map(Some)
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply sdk/src/derivation_path.rs:145]: `key` is an Option... but TIL about [`transpose`](https://doc.rust-lang.org/std/option/enum.Option.html#method.transpose), cool!

File: clap-utils/src/keypair.rs (hunk 1 above)
Comment (CriesofCarrots) [reply clap-utils/src/keypair.rs]: Great idea

File: sdk/src/derivation_path.rs [hunk 3]
```diff
@@ -120,6 +122,31 @@ impl DerivationPath {
             "".to_string()
         }
     }
+
+    // Only accepts single query string pair of type `key`
+    pub fn from_uri(uri: &URIReference<'_>) -> Result<Option<Self>, DerivationPathError> {
+        if let Some(query) = uri.query() {
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply sdk/src/derivation_path.rs:128]: In fact, it does. Test case and handling added: https://github.com/solana-labs/solana/pull/16933/commits/ab962226fc18a51baebc4e42be342e5eb25a8076

File: clap-utils/src/keypair.rs (hunk 1 above)
Comment (CriesofCarrots) [reply clap-utils/src/keypair.rs]: Done in: https://github.com/solana-labs/solana/pull/16933/commits/b354deaac92d6ce72be6fe14f24dff7d1dff8fa8
Wdyt?

File: clap-utils/src/keypair.rs [hunk 4]
```diff
@@ -196,9 +198,13 @@ pub(crate) fn parse_signer_source<S: AsRef<str>>(
                     ASK_KEYWORD => Ok(SignerSource::new(SignerSourceKind::Ask)),
                     _ => match Pubkey::from_str(source) {
                         Ok(pubkey) => Ok(SignerSource::new(SignerSourceKind::Pubkey(pubkey))),
-                        Err(_) => Ok(SignerSource::new(SignerSourceKind::Filepath(
-                            source.to_string(),
-                        ))),
+                        Err(_) => std::fs::OpenOptions::new()
```
Comment (t-nelson) [inline clap-utils/src/keypair.rs:201]: NIT: `std::fs::metadata()` is a little more direct

File: clap-utils/src/keypair.rs [hunk 5]
```diff
@@ -568,13 +591,13 @@ mod tests {
...
+        // Catchall into SignerSource::Filepath fails
+        let junk = "sometextthatisnotapubkeyorfile".to_string();
         assert!(Pubkey::from_str(&junk).is_err());
-        assert!(matches!(parse_signer_source(&junk).unwrap(), SignerSource {
-                kind: SignerSourceKind::Filepath(j),
-                derivation_path: None,
-            } if j == junk));
+        assert!(matches!(
+            parse_signer_source(&junk),
+            Err(SignerSourceError::IoError(_))
```
Comment (t-nelson) [inline clap-utils/src/keypair.rs:599]: :partying_face:

**Review by t-nelson — DISMISSED** [review]
Thanks! LGTM

**Pull request merged**